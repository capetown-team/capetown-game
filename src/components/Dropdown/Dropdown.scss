@import '../../styles/layout/colors';
@import '../../styles/layout/variables';

.dropdown {
  position: relative;
  margin-left: 5em;
  z-index: 3;

  &__panel {
    display: flex;
    align-items: center;
    cursor: pointer;
    padding: 0 20px;
    height: $headerHeight;
    border-radius: 5px;
    transition: background-color .3s;

    &_avatar {
      height: $headerHeight;
    }

    &_active {
      background-color: var(--background-secondary);
    }

    &:hover {
      background-color: var(--background-secondary);
    }
  }

  &__title {
    font-size: 1.3em;
    font-weight: bold;
    text-transform: capitalize;
    padding-right: 1em;
  }

  &__avatar {
    position: relative;
    overflow: hidden;
    width: 40px;
    height: 40px;
    border-radius: 50%;
    background-color: var(--background);
    background-repeat: no-repeat;
    background-image: url(data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNDAiIGhlaWdodD0iNDAiIHZpZXdCb3g9IjAgMCA0MCA0MCIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4NCjxwYXRoIGZpbGwtcnVsZT0iZXZlbm9kZCIgY2xpcC1ydWxlPSJldmVub2RkIiBkPSJNMzYgMkg0QzIuODk1IDIgMiAyLjg5NTQgMiA0VjI1LjI2NjdMMTQuNjU1IDIyLjMxMzlDMTUuNTQ5IDIyLjEwNTMgMTYuNDY0IDIyIDE3LjM4MSAyMkgyMi42MTlDMjMuNTM2IDIyIDI0LjQ1MSAyMi4xMDUzIDI1LjM0NSAyMi4zMTM5TDM4IDI1LjI2NjdWNEMzOCAyLjg5NTQgMzcuMTA1IDIgMzYgMlpNNCAwQzEuNzkxIDAgMCAxLjc5MDkgMCA0VjM2QzAgMzguMjA5MSAxLjc5MSA0MCA0IDQwSDM2QzM4LjIwOSA0MCA0MCAzOC4yMDkxIDQwIDM2VjRDNDAgMS43OTA5IDM4LjIwOSAwIDM2IDBINFpNMTAuOTA5IDE0LjU0NTVDMTIuOTE3IDE0LjU0NTUgMTQuNTQ1IDEyLjkxNzQgMTQuNTQ1IDEwLjkwOTFDMTQuNTQ1IDguOTAwOCAxMi45MTcgNy4yNzI3IDEwLjkwOSA3LjI3MjdDOC45MDEgNy4yNzI3IDcuMjcyOTkgOC45MDA4IDcuMjcyOTkgMTAuOTA5MUM3LjI3Mjk5IDEyLjkxNzQgOC45MDEgMTQuNTQ1NSAxMC45MDkgMTQuNTQ1NVoiIGZpbGw9IiNDRENEQ0QiLz4NCjwvc3ZnPg0K);
    background-size: 25px 25px;
    background-position: 50% 50%;
  }

  &__img {
    height: 100%;
    width: auto;
    position: absolute;
    left: 50%;
    transform: translateX(-50%);
  }

  &__arrow {
    background-image: url(data:image/svg+xml;charset=utf-8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20fill%3D%22none%22%20height%3D%228%22%20width%3D%2212%22%3E%3Cpath%20clip-rule%3D%22evenodd%22%20d%3D%22M2.156%202.295a.75.75%200%20011.051-.137L6%204.306l2.793-2.148a.75.75%200%2001.914%201.189l-3.25%202.5a.75.75%200%2001-.914%200l-3.25-2.5a.75.75%200%2001-.137-1.052z%22%20fill%3D%22%23aeb7c2%22%20fill-rule%3D%22evenodd%22%2F%3E%3C%2Fsvg%3E);
    background-position: center;
    background-size: cover;
    background-repeat: no-repeat;
    margin-top: 0;
    width: 12px;
    height: 8px;
    margin-left: 7px;
  }

  &__nav {
    display: none;
    position: absolute;
    top: $headerHeight;
    left: 0;
    right: 0;
    border-radius: .6rem;
    box-shadow: 0 2px 6px rgba(0, 0, 0, 0.2);
    background-color: var(--background-secondary);

    &_active {
      display: block;
    }
  }

  &__list {
    padding: 15px 10px;
    cursor: pointer;

    &:hover {
      background-color: rgba($light-blue, .1);
    }
  }
}
