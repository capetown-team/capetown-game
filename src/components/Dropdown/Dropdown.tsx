import React, { FC, memo, useState, useCallback } from 'react';
import { useHistory } from 'react-router-dom';
import block from 'bem-cn-lite';

import { baseUrl } from '@/constants';
import { ClickType } from '@/types.d';

import './Dropdown.scss';

const b = block('dropdown');

export type DropNavType = {
  name: string;
  redirect?: string;
  handler?: ClickType;
};

type Props = {
  name: string;
  avatar: string | null;
  nav: DropNavType[];
};

const Dropdown: FC<Props> = ({ name, avatar, nav }) => {
  const [isShow, setShow] = useState(false);
  const history = useHistory();

  const handlerToggleNav = useCallback(() => {
    setShow(!isShow);
  }, [isShow]);

  const handlerHideNav = useCallback(() => {
    setShow(false);
  }, []);

  const handlerRedirect = useCallback(
    (name) => {
      history.replace(name);
      setShow(false);
    },
    [history]
  );

  return (
    <div onMouseLeave={handlerHideNav} className={b()}>
      <div
        role="presentation"
        className={b('panel', { avatar: true, active: isShow })}
        onClick={handlerToggleNav}
      >
        <div className={b('title')}>{name}</div>
        <div className={b('avatar')}>
          <img
            className={b('img')}
            src={avatar ? `${baseUrl}/api/v2/resources/${avatar}` : ''}
            alt=""
          />
        </div>
        <div className={b('arrow')} />
      </div>

      <ul className={b('nav', { active: isShow })}>
        {nav.map(({ name, redirect, handler }) => {
          return redirect ? (
            <li
              key={name}
              role="presentation"
              className={b('list')}
              onClick={() => handlerRedirect(redirect)}
            >
              {name}
            </li>
          ) : (
            <li
              key={name}
              role="presentation"
              className={b('list')}
              onClick={handler}
            >
              {name}
            </li>
          );
        })}
      </ul>
    </div>
  );
};

const WrappedDropdown = memo(Dropdown);

export { WrappedDropdown as Dropdown };
