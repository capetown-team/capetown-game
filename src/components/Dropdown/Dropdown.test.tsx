import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { DropNavType, Dropdown } from '@/components/Dropdown/Dropdown';

describe('<Authorization />', () => {
  it('should renders correct <Dropdown />', () => {
    const dropLists: DropNavType[] = [
      {
        redirect: '/profile',
        name: 'Профиль'
      },
      {
        redirect: '/exit',
        name: 'Выйти'
      }
    ];

    const wrapper = shallow(
      <Dropdown name="Клава" avatar="test" nav={dropLists} />
    );

    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
