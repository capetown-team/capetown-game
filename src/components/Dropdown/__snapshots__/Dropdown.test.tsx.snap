// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Authorization /> should renders correct <Dropdown /> 1`] = `
<div
  className="dropdown"
  onMouseLeave={[Function]}
>
  <div
    className="dropdown__panel dropdown__panel_avatar"
    onClick={[Function]}
    role="presentation"
  >
    <div
      className="dropdown__title"
    >
      Клава
    </div>
    <div
      className="dropdown__avatar"
    >
      <img
        alt=""
        className="dropdown__img"
        src="https://ya-praktikum.tech/api/v2/resources/test"
      />
    </div>
    <div
      className="dropdown__arrow"
    />
  </div>
  <ul
    className="dropdown__nav"
  >
    <li
      className="dropdown__list"
      key="Профиль"
      onClick={[Function]}
      role="presentation"
    >
      Профиль
    </li>
    <li
      className="dropdown__list"
      key="Выйти"
      onClick={[Function]}
      role="presentation"
    >
      Выйти
    </li>
  </ul>
</div>
`;
